plugins/com.google.cloud.tools.eclipse.appengine.ui/src/com/google/cloud/tools/eclipse/appengine/ui/AppEngineLibrariesSelectorGroup.java

```diff
@@ -107,55 +104,205 @@ private static String getLibraryName(Library library) {
+                             new UpdateValueStrategy());
+    // library checkbox enablement model -> UI
+    bindingContext.bindValue(libraryButtonEnablement,
+                             new DependentLibrarySelected(getDisplayRealm(), selectedLibraries, library.getId(), false),
+                             new UpdateValueStrategy(UpdateValueStrategy.POLICY_NEVER),
+                             new UpdateValueStrategy());
+  }
+
+  public void dispose() {
+    if (bindingContext != null) {
+      bindingContext.dispose();
+    }
+  }
+
+  private Realm getDisplayRealm() {
+    return DisplayRealm.getRealm(parentContainer.getDisplay());
+  }
+
+  @VisibleForTesting
+  List<Button> getLibraryButtons() {
+    return libraryButtons;
+  }
+
+  /**
+   * Utility interface to return a value;
+   */
+  private static interface Getter<T> {
+    T get();
+  }
+
+  private static final class ButtonManuallySelected implements Getter<Boolean> {
+    private final Button libraryButton;
+
+    private ButtonManuallySelected(Button libraryButton) {
+      this.libraryButton = libraryButton;
+    }
+
+    /**
+     * Returns true if the checkbox associated with this instance was explicitly clicked by the user.
+     */
+    @Override
+    public Boolean get() {
+      return libraryButton.getData(BUTTON_MANUAL_SELECTION_KEY) != null;
     }
   }
 
-  private Button getButtonForLibraryId(String libraryId) {
-    for (Button button : libraryButtons) {
-      Library library = (Library) button.getData();
-      if (library.getId().equals(libraryId)) {
-        return button;
+  /**
+   * Tracks if the checkbox has been explicitly clicked by the user.
+   */
+  private static final class ManualSelectionTracker implements SelectionListener {
+    @Override
+    public void widgetSelected(SelectionEvent e) {
+      setManualSelection(e);
+    }
+
+    @Override
+    public void widgetDefaultSelected(SelectionEvent e) {
+      setManualSelection(e);
+    }
+
+    private void setManualSelection(SelectionEvent e) {
+      Button source = (Button)e.getSource();
+      if (e.getSource() instanceof Button && (source.getStyle() & SWT.CHECK) != 0) {
+        Button button = source;
+        button.setData(BUTTON_MANUAL_SELECTION_KEY, button.getSelection() ? new Object() : null);
       }
     }
-    return null;
   }
 
-  private Realm getDisplayRealm() {
-    return DisplayRealm.getRealm(parentDialog.getControl().getDisplay()); 
+  /**
+   * Returns a computed value based on whether the associated library is in a list or not. 
+   */
+  private static final class DependentLibrarySelected extends ComputedValue {
+    private String libraryId;
+    private IObservableList libraries;
+    private Getter<Boolean> condition;
+    private boolean selectedResult;
+
+    /**
+     * @param libraries the list of libraries to search
+     * @param libraryId the id of the library to be searched for
+     * @param foundResult return value if the library is found
```

Comment:
I am confused by this comment initially. I suggest to change it to `the value calculate() will return if the library is found` or something along the line.